### PR TITLE
gogio: [android] support AAR [WIP] 

### PR DESCRIPTION
**Pull request by inkeliz**
That handles `AAR` similar to `JAR`.

TODO:
- [x] gogio: Find AAR.
- [x] d8: Extract and merge JAR from AAR.
- [x] aapt: Generate assets from all AAR and combine.
- [x] manifest: Generate and merge all AndroidManifests. (Requires external tool!) 
- [ ] AIDL: Generate AIDL from AAR.
- [ ] lib: Merge all libs (.so) from AAR.
- [ ] gogio: detect and resolve dependencies/imports conflicts.

CC @whereswaldon

### CODE REVIEW

**Comment by gedw99** [comment]
any chance this can get reviewed ?

**Comment by eliasnaur** [comment]
I'd review, but the PR is marked as a draft.

**Comment by eliasnaur** [comment]
I'm sorry for missing this. GitHub apparently doesn't notify me when the PR status changes from draft to reviewable.

Some high level comments:

This PR seems to mix AAR support with schema support and still includes `[WIP]` in the title.

Adding AAR support to the `gogio` tool adds significant maintenance burden, because there are just so many things an .aar file can contain. Just like some Go users need to compile Go code with buildmode `c-archive` or `c-shared` perhaps it's ok to use `buldmode` `archive` for `gogio` and let users build the final Android app themselves?

To expand the second question: given a `gogio -buildmode archive` .aar file, is it possible for an external tool to merge that aar file with another, without knowing the intricacies of Go nor Gio? If so, that is a sign this PR does not belong in `gogio`.

**Comment by inkeliz** [comment]
> To expand the second question: given a gogio -buildmode archive .aar file, is it possible for an external tool to merge that aar file with another, without knowing the intricacies of Go nor Gio? If so, that is a sign this PR does not belong in gogio.

The challenge is manage Java and Go dependencies independently. For instance, some _gio-x_ packages use Fragment, _which is deprecated for while now_. Moving to [AndroidX Fragment](https://developer.android.com/jetpack/androidx/releases/fragment) requires an external .aar. If we opt for "just documenting it," everyone will be required to include it as a dependency in Maven/Gradle or to download the .aar file.  This issue extends to other features, like the Barcode Scanner, which is natively available on most platforms except Android. In the end, I can't use `go get gio-x/barcode` and `gogio -target android`. In my case, I need to use an external `.aar` for ["Sign in with Google"](https://github.com/gioui-plugins/gio-plugins/pull/58/commits).

This is more about "ease of use." As far as I remember, the goal behind "gogio" was to facilitate the creation of demonstrations and prototypes, not final builds. Maybe, the initial goal already drift a little bit. But, that can be a hint that some people prefer to use `gogio` instead of some dedicated per-OS tool, and it's good enough.

**Comment by ddkwork** [comment]
> I'm sorry for missing this. GitHub apparently doesn't notify me when the PR status changes from draft to reviewable.
> 
> Some high level comments:
> 
> This PR seems to mix AAR support with schema support and still includes `[WIP]` in the title.
> 
> Adding AAR support to the `gogio` tool adds significant maintenance burden, because there are just so many things an .aar file can contain. Just like some Go users need to compile Go code with buildmode `c-archive` or `c-shared` perhaps it's ok to use `buldmode` `archive` for `gogio` and let users build the final Android app themselves?
> 
> To expand the second question: given a `gogio -buildmode archive` .aar file, is it possible for an external tool to merge that aar file with another, without knowing the intricacies of Go nor Gio? If so, that is a sign this PR does not belong in `gogio`.

I think this PR can expand the potential of Gio. I successfully ran webview on an Android phone using his plugin, but more permissions need to be implanted into the XML during the compilation phase. Accessing webview using Android browser is normal, APK requires more permissions.

**Comment by eliasnaur** [comment]
> > To expand the second question: given a gogio -buildmode archive .aar file, is it possible for an external tool to merge that aar file with another, without knowing the intricacies of Go nor Gio? If so, that is a sign this PR does not belong in gogio.
> 
> The challenge is manage Java and Go dependencies independently. For instance, some _gio-x_ packages use Fragment, _which is deprecated for while now_. Moving to [AndroidX Fragment](https://developer.android.com/jetpack/androidx/releases/fragment) requires an external .aar. If we opt for "just documenting it," everyone will be required to include it as a dependency in Maven/Gradle or manually download the .aar file and glue it. This issue extends to other features, like the Barcode Scanner, which is natively available on most platforms except Android. In the end, I can't use `go get gio-x/barcode` and `gogio -target android`. In my case, I need to use an external `.aar` for ["Sign in with Google"](https://github.com/gioui-plugins/gio-plugins/pull/58/commits).
> 
> This is more about "ease of use." As far as I remember, the goal behind "gogio" was to facilitate the creation of demonstrations and prototypes, not final builds. Maybe, the initial goal already drift a little bit. But, that can be a hint that some people prefer to use `gogio` instead of some dedicated per-OS tool, and it's good enough.

Good points.

I understand the need for easier packaging of Android apps (god knows I despise the official ones), but because of our limited resources, I think it's important to focus our efforts. That's why I model `gogio`'s behaviour like the `go` tool: it can build "executables" for the common, straightforward cases and leaves more complicated cases for `-buildmode c-archive` or `-buildmode c-shared`.

In other words, I believe `gogio -buildmode archive` is what should package the Go/Gio specific code and resources in a way suitable for Android standard tools, and nothing else.

Now, there's nothing stopping a higher-level tool that can take .AAR files/gradle projects/... along with metadata to produce a final (signed?) .apk. The advantages of a separate tool are:

1. Smaller maintenance burden on the Gio developers.
2. Useful for any project: Gio, Fyne and even apps that don't use Go at all.
3. Because of 2. the potential community is much larger and thus easier to find developers for.

**Comment by eliasnaur** [comment]
Unless you have objections to my reasons for delegating AAR packaging to a separate tool, I'm inclined to close this. WDYT?
